src/bit_str.jl

```diff
@@ -1,124 +1,98 @@
-
-"""
-    bit(string)
-
-Create a [`BitStr`](@ref) with given string of bits. See also [`@bit_str`](@ref).
-"""
-bit(x::String) = BitStr(x)
-
-"""
-    bit(x[; len=ndigits(x, base=2)])
-
-Create a [`BitStr`](@ref) accroding to integer `x` to given length `len`.
-"""
-bit(x::Integer; len=ndigits(x, base=2)) = BitStr(x, len)
-
-"""
-    bit(;len)
-
-Lazy curried version of [`bit`](@ref).
-"""
-bit(;len) = x -> bit(x; len=len)
-
-"""
-    bit_literal(xs...)
+for op in [:<, :>, :(<=), :(>=)]
+    @eval Base.$op(a::T, b::T) where T<:BitStr = Base.$op(Integer(a),Integer(b))
+end
 
-Create a [`BitStr`](@ref) by input bits `xs`.
+for op in [:(==)]
+    @eval Base.$op(a::T, b::Number) where T<:BitStr = Base.$op(Integer(a),b)
+    @eval Base.$op(a::Number, b::T) where T<:BitStr = Base.$op(a,Integer(b))
+    @eval Base.$op(a::T, b::T) where T<:BitStr = Base.$op(Integer(a),Integer(b))
+end
+for op in [:count_ones, :count_zeros, :leading_ones, :leading_zeros]
+    @eval Base.$op(a::BitStr) = Base.$op(Integer(a))
+end
 
-# Example
+# Note: the transitivity of == is not satisfied here.
+Base.:(==)(lhs::BitStr, rhs::BitStr) = false
+Base.isapprox(a::BitStr, b::Number; kwargs...) = Base.isapprox(Integer(a),b; kwargs...)
+Base.isapprox(a::Number, b::BitStr; kwargs...) = Base.isapprox(a,Integer(b); kwargs...)
+Base.isapprox(lhs::BitStr, rhs::BitStr; kwargs...) = false
+Base.isapprox(a::T, b::T; kwargs...) where T<:BitStr = Base.isapprox(Integer(a),Integer(b); kwargs...)
 
-```jldoctest
-julia> bit_literal(1, 0, 1, 0, 1, 1)
-110101 (53)
-```
-"""
-bit_literal(xs...) = bit_literal(xs)
-function bit_literal(xs::NTuple{N, Int}) where N
-    val = zero(int(N))
-    for k in 1:N
-        xs[k] == 0 || xs[k] == 1 || error("expect 0 or 1, got $(xs[k])")
-        val += xs[k] << (k - 1)
-    end
-    return BitStr(val, N)
-end
+# Note: it is a bit confusing, with x::BitStr == y::Integer,
+# they behave different when used for indexing.
+Base.to_index(x::BitStr) = Integer(x) + 1
+Base.to_index(x::UnitRange{<:BitStr}) = Integer(x.start) + 1:Integer(x.stop) + 1
 
 # use system interface
-Base.to_index(x::BitStr) = Int(x.val) + 1
 Base.checkindex(::Type{Bool}, inds::AbstractUnitRange, i::BitStr) =
     checkindex(Bool, inds, Base.to_index(i))
-Base.length(bits::BitStr{<:Integer, N}) where N = N
+Base.length(bits::BitStr{N}) where N = N
 Base.lastindex(bits::BitStr) = length(bits)
 
-"""
-    to_location(x)
-
-Convert bit configuration `x` to an index.
-
-# Example
-
-```jldoctest
-julia> to_location(1)
-2
-
-julia> to_location(bit"111")
-111 (7)
-```
-"""
-to_location(x) = error("expect an integer or @bit_str, got $(typeof(x))")
-to_location(x::Integer) = x + 1
-to_location(x::BitStr) = x
-
+Base.typemax(::Type{BitStr{N,T}}) where {N,T} = BitStr{N,T}(1<<N-1)
+Base.typemin(::Type{BitStr{N,T}}) where {N,T} = BitStr{N,T}(0)
+Base.typemax(::BitStr{N,T}) where {N,T} = BitStr{N,T}(1<<N-1)
+Base.typemin(::BitStr{N,T}) where {N,T} = BitStr{N,T}(0)
```

Comment:
if `typemin` is `0` the storage type should be `UInt64`.

Reply:
If the storage type if `UInt64`, we will have more trouble in arithmatics.

Here, both `typemax` and `typemin` are not related to storage. I don't know how to check the bounds.